batched evaluation

dkrako opened the pull request:
### Description
- This PR implements the classes `BatchedMetric` and `BatchedPerturbationMetric`
- This is a rewrite of this PR: #87 
- It solves the following issue: #80 
- This will speed up a lot of metrics.

### Implemented changes
- Insert a description of the changes implemented in the pull request.
  - [x] Add `BatchedMetric` and `BatchedPerturbationMetric`.
  - [x] Add support for additional kwargs in `Metric`, if these are present in `evaluate_instance` signature.
  - [x]  Apply signature change to all classes inheriting from `PerturbationMetric` to the use improved kwargs-passing.
  - [x] Transform `InputInvariance` superclass from `PerturbationMetric` to `BatchedPerturbationMetric`.

### Minimum acceptance criteria
- [x] Pass all tests.

leanderweber commented on quantus/metrics/robustness/max_sensitivity.py:
```diff
@@ -258,75 +260,105 @@ def __call__(
+        nr_samples: int = None,
+        similarity_func: Callable = None,
+        norm_numerator: Callable = None,
+        norm_denominator: Callable = None,
```
not sure about making all of these arguments to the evaluate_batch / evaluate_instance method...
since they are parameters that are supposed to define the metric, and be set (only) in __init__.

but this is not that important for now. I think we should discuss which hyperparameters go where again anyway sometime soon, especially wrt normalisation/abs/aggregation.

dkrako replied:
yes that's also a point i wanted to discuss.
i personally prefer to have these additional metric variables as method arguments instead of attributes.
this way i can already gauge from the method signature what will be actually used during computation.
but that's just my preferred way of doing this, and attributes will work as well.

as this is more or less hidden to the user who basically uses only `__init__()` and `__call__()`, the discussion to handle this consistently could be deferred to a point after the merge (and create an issue to not forget).

dkrako replied:
we should also discuss which of the methods of a `Metric` should be marked as private.

Maybe `general_preprocess()`, `custom_preprocess()`, `custom_postprocess()`, `evaluate_instance()` and so on should be renamed to `__general_preprocess()`, `__custom_preprocess()`, `__custom_postprocess()`, `__evaluate_instance()` respectively.

This way the API html doc-page of each specific metric implementation won't be crowded by methods the actual users don't care about.
The base metric implementation could still include the documentation for these private methods and will be then the goto-reference for developers of new metrics.

annahedstroem replied:
Also, adding the variables as arguments to the evaluate_batch method just duplicates a lot of documentation, cause everywhere we pass it, it has to be re-documented.

annahedstroem replied:
Daniel can we, for this PR, use the self.similarity_func instead of passing similarity_func? Unless there is a big benefit of the approach that I cannot see

annahedstroem replied:
Agree on this: _"Maybe general_preprocess(), custom_preprocess(), custom_postprocess(), evaluate_instance() and so on should be renamed to __general_preprocess(), __custom_preprocess(), __custom_postprocess(), __evaluate_instance() respectively."_

but scope for a separate PR

dkrako replied:
I see your points and have now removed all the stuff regarding additional arguments in the evaluate methods.

leanderweber reviewed (approved):
two minor comments, but otherwise seems good to me

annahedstroem commented on quantus/helpers/perturb_func.py:
```diff
@@ -27,6 +27,32 @@
+        inplace: bool = False,
+        **kwargs,
+) -> Union[np.ndarray, None]:
+    """ perturbation of complete batch """
```
Let's me consistent with the docstrings, following numpydoc https://numpydoc.readthedocs.io/en/latest/format.html

annahedstroem commented on quantus/helpers/perturb_func.py:
```diff
@@ -27,6 +27,32 @@
 )
 
 
+def perturb_batch(
```
Also let's comment this function more.

annahedstroem commented on quantus/metrics/robustness/avg_sensitivity.py:
```diff
@@ -110,13 +111,27 @@ def __init__(
             normalise_func_kwargs=normalise_func_kwargs,
             perturb_func=perturb_func,
             perturb_func_kwargs=perturb_func_kwargs,
+            similarity_func=similarity_func,
```
This is applicable on all metrics

annahedstroem replied:
Let me know if you want to sync on it during a short meeting?

annahedstroem commented on quantus/metrics/robustness/continuity.py:
```diff
@@ -294,12 +300,12 @@ def evaluate_instance(
 
             # Generate explanation based on perturbed input x.
             dx_step = (step + 1) * self.dx
-            x_perturbed = self.perturb_func(
```
Also, why not re-using the class attributes for perturb_func and perturb_func_kwargs, so keeping the self?

dkrako replied:
done

annahedstroem commented on quantus/metrics/base_batched.py:
```diff
@@ -0,0 +1,404 @@
+        display_progressbar (boolean): Indicates whether a tqdm-progress-bar is printed.
+
+        """
+        # We need to do this separately here, as super().__init__() overwrites these
```
What does "do this" mean in "We need to do this separately here"

dkrako replied:
i have removed this custom kwargs adding as we now dont change the evaulate method signature.

annahedstroem commented on quantus/metrics/base_batched.py:
```diff
@@ -0,0 +1,404 @@
+            display_progressbar=self.display_progressbar,
+        )
+
+        # TODO: initialize correct length of last results
```
Are these completed or still left to do? Is it necessary to fix this to merge/ production

dkrako replied:
no not necessary

dkrako replied:
removed the TODO

annahedstroem commented on quantus/metrics/base_batched.py:
```diff
@@ -0,0 +1,404 @@
+        **kwargs,
+    ):
+        """
+        Initialise the Metric base class.
```
BatchedMetric

annahedstroem commented on quantus/metrics/base_batched.py:
```diff
@@ -0,0 +1,404 @@
+        """
+        Initialise the Metric base class.
+
+        Each of the defined metrics in Quantus, inherits from Metric base class.
```
BatchedMetric

annahedstroem commented on quantus/metrics/base_batched.py:
```diff
@@ -0,0 +1,404 @@
+
+    @staticmethod
+    def get_number_of_batches(n_instances: int, batch_size: int):
+        return math.ceil(n_instances / batch_size)
```
We need to do the numpydoc also here! I'll make a suggestion :)

annahedstroem commented on quantus/metrics/base_batched.py:
```diff
@@ -0,0 +1,404 @@
+            if not isinstance(value, (Sequence, np.ndarray)) or isinstance(value, str):
+                single_value_kwargs[key] = value
+
+            # Check if sequence has correct length.
```
Do we need to fix this right now or can we separate this into another issue? I think it would be great if we could scope this outside.

annahedstroem commented on quantus/metrics/base_batched.py:
```diff
@@ -0,0 +1,404 @@
+            data: Dict[str, Any],
+            batch_size: int,
+            display_progressbar: bool = False,
+    ):
```
same :)

annahedstroem replied:
and let's add some comments to this function in general, especially in the end

dkrako replied:
done

annahedstroem commented on quantus/metrics/base_batched.py:
```diff
@@ -0,0 +1,404 @@
+            yield {**batch, **single_value_kwargs}
+
+    def evaluate_instance(self, **kwargs) -> Any:
+        raise NotImplementedError('evaluate_instance() not implemented for BatchedMetric')
```
same :)

annahedstroem commented on quantus/metrics/base_batched.py:
```diff
@@ -0,0 +1,404 @@
+        batch_size: int = 64,
+        **kwargs,
+    ) -> Union[int, float, list, dict, None]:
+        return super().__call__(
```
documentation for the remaining functions

dkrako replied:
done

annahedstroem reviewed (commented):
Thanks so much @dkrako! Some things:
- Re-use class attributes for now?
- Adding numpydoc style documentation everywhere (this is super key)
- Separating #TODOs if possible to other Issues, so that we can proceed with a relatively fast merge so that we can cont. with the other PRs

dkrako commented on quantus/metrics/axiomatic/completeness.py:
```diff
@@ -247,9 +247,11 @@ def evaluate_instance(
+        y: np.ndarray = None,
+        a: np.ndarray = None,
+        s: np.ndarray = None,
+        perturb_func: Callable = None,
```
i have removed it and replaced it by attributes

dkrako commented on quantus/metrics/axiomatic/input_invariance.py:
```diff
@@ -222,60 +224,82 @@ def __call__(
+        perturb_func_kwargs: Dict = None,
+    ) -> np.ndarray:
         """
         Evaluate instance gets model and data for a single instance as input and returns the evaluation result.
```
done

dkrako commented on quantus/metrics/base_batched.py:
```diff
@@ -0,0 +1,404 @@
+            a_batch: np.ndarray,
+            s_batch: Optional[np.ndarray] = None,
+    ):
+        raise NotImplementedError()
```
done

dkrako commented on quantus/metrics/base_batched.py:
```diff
@@ -0,0 +1,404 @@
+        if kwargs:
+            evaluate_kwarg_names = inspect.getfullargspec(self.evaluate_batch).args
+            for key, value in list(kwargs.items()):
+                if key in evaluate_kwarg_names or re.sub('_batch', '', key) in evaluate_kwarg_names:
```
not relevant anymore, but added short comment to Metric.get_instance_iterator()

dkrako commented:
I think I have addressed the issues!

- [x] removed all the additional argument stuff from the evaluate methods.
- [x] added and adjusted docstrings
- [x] removed or implemented TODOs

annahedstroem commented on quantus/metrics/base.py:
```diff
@@ -610,7 +610,8 @@ def get_instance_iterator(self, data: Dict[str, Any]):
                 data[key] = [value for _ in range(n_instances)]
 
-        # We create a list of dictionaries where each dictionary holds all data for a single instance
+        # We create a list of dictionaries where each dictionary holds all data for a single instance.
```
Super clear. Thank you.

annahedstroem reviewed (commented):
Great. Will push tiny tweaks, then ready for merge 🎉
